ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionLogUpdatesConversions.scala

```diff
@@ -10,66 +10,82 @@ import com.daml.ledger.api.v1.transaction.{
+        val flatTransactionEvents = transactionEvents.collect {
+          case createdEvent: TxUpdate.CreatedEvent => createdEvent
+          case exercisedEvent: TxUpdate.ExercisedEvent if exercisedEvent.consuming =>
+            exercisedEvent
+        }
+        val filteredFlatEvents = flatTransactionEvents
+          .filter(flatTransactionPredicate(wildcardParties, templateSpecificParties))
+        val commandIdO = filteredFlatEvents.collectFirst {
+          case event if event.commandId.nonEmpty => event.commandId
+        }
+        // Allows emitting flat transactions with no events, a use-case needed
+        // for the functioning of Daml triggers.
+        // (more details in https://github.com/digital-asset/daml/issues/6975)
+        Option.when(filteredFlatEvents.nonEmpty || commandIdO.nonEmpty)(
+          transaction.copy(events = filteredFlatEvents)
+        )
+      }
+
+    def toApiTransaction(
         filter: FilterRelation,
-        wildcardParties: Set[String],
-        templateSpecificParties: Map[Identifier, Set[String]],
         verbose: Boolean,
         lfValueTranslation: LfValueTranslation,
     )(implicit
         loggingContext: LoggingContext,
         executionContext: ExecutionContext,
-    ): Future[Option[FlatTransaction]] = {
-      val transactionEvents = tx.events
-      val filtered = transactionEvents.filter(
-        FlatTransactionPredicate(_, wildcardParties, templateSpecificParties)
-      )
-
-      filtered.headOption
-        .map { first =>
-          val events = removeTransient(filtered)
-          val requestingParties = filter.keySet.map(_.toString)
-
-          Future
-            .traverse(events)(toFlatEvent(_, requestingParties, verbose, lfValueTranslation))
-            .map(_.collect { case Some(value) => value })
-            .map { flatEvents =>
-              // Allows emitting flat transactions with no events, a use-case needed
-              // for the functioning of Daml triggers.
-              // (more details in https://github.com/digital-asset/daml/issues/6975)
-              if (flatEvents.nonEmpty || first.commandId.nonEmpty) {
-                Some(
-                  FlatTransaction(
-                    transactionId = first.transactionId,
-                    commandId = getCommandId(events, requestingParties),
-                    workflowId = first.workflowId,
-                    effectiveAt = Some(timestampToTimestamp(first.ledgerEffectiveTime)),
-                    events = flatEvents,
-                    offset = ApiOffset.toApiString(tx.offset),
-                  )
-                )
-              } else None
-            }
-        }
-        .getOrElse(Future.successful(None))
+    ): ToApi[GetTransactionsResponse] = { transaction =>
+      val nonTransient = removeTransient(transaction.events)
+      val requestingParties = filter.keySet
+      Future
+        .traverse(nonTransient)(toFlatEvent(_, requestingParties, verbose, lfValueTranslation))
```

Comment:
just by doing the Future traverse here we will sync execute all toFlatEvent future factory functions, which unfortunately boil down to sync Engine call if verbose mode is on.
I advise caution executing lfValueTranslation code therefore, and wrapping the Future.traverse in a Future. An example for that you can see in TransactionReader.getActiveContracts:
```
          future = Future(
            Future.traverse(rawResult)(
              deserializeEntry(verbose)
            )
          ).flatMap(identity),
```
In the acs case it is more important because there deserialization logic is running async, bu since the engine call in this example I think it is still a good idea for being careful.

(we have pending tasks to fix the root cause here, but it was not scheduled/implemented yet)

Reply:
until the toFlatEvent future factories are sync, they are also blocking in traverse, so as non-intuitively: this wont have non-limited parallelism (all entries executing in parallel) async, but rather sequential blocking sync call (upto the async part in the toFlatEvent method).
With wrapping this to a Future we would have that sequential (parallelism 1) async call.

Reply:
Good remarks. I'm leaving this change to the subsequent PR that is meant to address exactly these performance concerns (https://github.com/digital-asset/daml/pull/13880) in order to reduce the impact of this one since it was already validated from a performance POV.